### PR TITLE
Dilpreet | Stage | Creation of sales, purchase, credit note and debit note entries and voucher from the journal module

### PR DESCRIPTION
**Pull request by dilpreetsinghofficial**
https://app.clickup.com/t/86cv5uhtu

### CODE REVIEW

File: apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts [hunk 1]
```diff
@@ -766,21 +1177,19 @@ export class AccountAsVoucherComponent implements OnInit, OnDestroy, AfterViewIn
...
-                    this.openChequeDetailForm();
-                }
+            if (response?.body && (response?.body?.currency?.code || this.activeCompany?.baseCurrency) === this.activeCompany?.baseCurrency) {
                 let idx = this.selectedIdx;
                 transactionsFormArray = this.journalVoucherForm.get('transactions') as FormArray;
                 transactionAtIndex = transactionsFormArray.at(idx) as FormGroup;
-
+                if (acc && acc.parentGroups.find((pg) => pg?.uniqueName === 'bankaccounts')) {
+                    this.openChequeDetailForm();
+                } else {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts]: Blank else

File: apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts [hunk 2]
```diff
@@ -851,9 +1305,35 @@ export class AccountAsVoucherComponent implements OnInit, OnDestroy, AfterViewIn
...
             }
+            this.checkSalesEntry();
+        });
+    }
+    /**
+     * This will be use for check sales entry
+     *
+     * @memberof AccountAsVoucherComponent
+     */
+    public checkSalesEntry() {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts]:
```suggestion
    public checkSalesEntry(): void {
```

File: apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts [hunk 3]
```diff
@@ -921,45 +1374,41 @@ export class AccountAsVoucherComponent implements OnInit, OnDestroy, AfterViewIn
...
-                }
-            } else if (this.totalDebitAmount < this.totalCreditAmount || (this.totalCreditAmount === 0 && this.totalDebitAmount === 0)) {
+    /**
+     * This will be use for check voucher type new entries
+     *
+     * @param {*} currentVoucher
+     * @param {*} voucherType
+     * @memberof AccountAsVoucherComponent
+     */
+    public checkVoucherTypeNewEntries(currentVoucher: any, voucherType: any, transactionObj?: any): void {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts:1400]:
```suggestion
    public checkVoucherTypeNewEntries(currentVoucher: any, voucherType: string, transactionObj?: any): void {
```

File: apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts [hunk 4]
```diff
@@ -1002,15 +1446,44 @@ export class AccountAsVoucherComponent implements OnInit, OnDestroy, AfterViewIn
...
+
+        let totalCredit = 0;
+        let totalDebit = 0;
+
+        (this.journalVoucherForm.get('transactions') as FormArray).controls?.forEach((control: FormGroup) => {
+            if (control.get('type').value.toLowerCase() === 'to') {
+                if (!control.get('isDiscountApplied')?.value) {
+                    totalCredit += control.get('amount').value;
+                }
+            } else {
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts]:
```suggestion
            } else if (!control.get('isDiscountApplied')?.value) {
```

File: apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts [hunk 5]
```diff
@@ -2020,12 +2533,196 @@ export class AccountAsVoucherComponent implements OnInit, OnDestroy, AfterViewIn
         }
     }
 
+    /**
+     * Keydown handler event
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/journal-voucher/voucher/voucher.component.ts]:
```suggestion
     * Key up/down event handler
```

File: apps/web-giddh/src/app/accounting/ng-virtual-list/virtual-list.component.ts [hunk 6]
```diff
@@ -290,16 +290,22 @@ export class AVShSelectComponent implements ControlValueAccessor, OnInit, AfterV
         this.cdRef.markForCheck();
     }
 
-    public keydownUp(event): void {
+    /**
+     * This will be use for keyboard handler
```
Comment (dvCodeWorld) [inline apps/web-giddh/src/app/accounting/ng-virtual-list/virtual-list.component.ts]:
```suggestion
     * This will be use for keyboard event handler
```

**Review by dvCodeWorld — CHANGES_REQUESTED** [review]

**Review by dvCodeWorld — APPROVED** [review]

**Review by ravinderwalkover — APPROVED** [review]

**Pull request merged**